jumanji/environments/routing/rware/env.py

```diff
@@ -0,0 +1,527 @@
+            agent_id: the id of the agent performing the action.
+        Returns:
+            the updated warehouse floor grid array, agents and shelves.
+        """
+        agent = tree_slice(agents, agent_id)
+        is_highway = self.highways[agent.position.x, agent.position.y]
+        grid, agents, shelves = jax.lax.cond(
+            jnp.equal(action, Action.FORWARD.value),
+            utils.set_new_position_after_forward,
+            utils.set_new_direction_after_turn,
+            grid,
+            agents,
+            shelves,
+            action,
+            agent_id,
+            is_highway,
+        )
+
+        return grid, agents, shelves
+
+    def _update_reward_and_request_queue(
+        self,
+        key: chex.PRNGKey,
+        reward: chex.Array,
+        request_queue: chex.Array,
+        grid: chex.Array,
+        shelves: chex.Array,
+        goal: chex.Array,
+    ) -> Tuple[chex.PRNGKey, int, chex.Array, chex.Array]:
+        """Check if a shelf has been delivered successfully to a goal state,
+        if so reward the agents and update the request queue: removing the ID
+        of the delivered shelf and replacing it with a new shelf ID.
+
+        Args:
+            key: a pseudorandom number generator key.
+            reward: the array of shared reward for each agent.
+            request_queue: the queue of requested shelves.
+            grid: the warehouse floor grid array.
+            shelves: a pytree of Shelf type containing shelves information.
+            goal: array of goal positions.
+        Returns:
+            a random key, updated reward, request queue and shelves.
+        """
+        x, y = goal
+        shelf_id = grid[_SHELVES, x, y]
+
+        def reward_and_update_request_queue_if_shelf_in_goal(
+            key: chex.PRNGKey,
+            reward: jnp.int32,
+            request_queue: chex.Array,
+            shelves: chex.Array,
+            shelf_id: int,
+        ) -> Tuple[chex.PRNGKey, int, chex.Array, chex.Array]:
+            "Reward the agents and update the request queue."
+
+            # remove from queue and replace it
+            key, request_key = jax.random.split(key)
+
+            not_in_queue = jnp.setdiff1d(
+                self.shelf_ids,
+                request_queue,
+                size=self.not_in_queue_size,
+            )
+            new_request_id = jax.random.choice(
+                request_key,
+                not_in_queue,
+                replace=False,
+            )
+            replace_index = jnp.argwhere(jnp.equal(request_queue, shelf_id - 1), size=1)
+            request_queue = request_queue.at[replace_index].set(new_request_id)
+
+            # also reward the agents
+            reward += 1.0
+
+            # update requested shelf
+            shelves = utils.update_shelf(shelves, shelf_id - 1, "is_requested", 0)
+            shelves = utils.update_shelf(shelves, new_request_id, "is_requested", 1)
```

Comment:
This is also where the `**kwargs` would be nice, then you'd only have to call this method once

Reply:
Not sure this would work, since the argument name would be the same but have two different values. Therefore leaving this as is for now, but if I misunderstood, please feel free to let me know. Of course, in other cases `**kwargs` makes sense, i.e. updating to different attributes.